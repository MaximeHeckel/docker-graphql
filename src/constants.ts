const baseURL =
  process.env.BASE_URL || "http+unix://%2Fvar%2Frun%2Fdocker.sock";

const port = process.env.PORT || "4000";

export { baseURL, port };
