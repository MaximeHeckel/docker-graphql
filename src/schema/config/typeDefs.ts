import { gql } from 'apollo-server';

const typeDefs = gql`
  extend type Query {
    configs: [Config!]!
    config(id: ID): Config!
  }

  type ConfigSpec {
    Name: String!
  }

  type Config {
    ID: ID!
    Version: VersionType
    CreatedAt: String!
    UpdatedAt: String!
    Spec: ConfigSpec
  }
`;

export default typeDefs;
