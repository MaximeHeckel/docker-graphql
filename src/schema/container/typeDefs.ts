import { gql } from "apollo-server";

const typeDefs = gql`
  extend type Query {
    containers: [Container!]!
    container(id: ID!): Container!
  }
  type Container {
    Id: String!
    Command: String!
    Image: String!
    Names: [String!]!
    State: String!
    Status: String!
  }
`;

export default typeDefs;
