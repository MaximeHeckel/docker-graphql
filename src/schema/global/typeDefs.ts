import { gql } from "apollo-server";

export const Query = gql`
  type Query
`;

const global = [Query];

export default global;
