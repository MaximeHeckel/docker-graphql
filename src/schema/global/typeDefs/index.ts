import { Query } from './entrypoint';
import scalars from './scalars';

const global = [Query, scalars];

export default global;
