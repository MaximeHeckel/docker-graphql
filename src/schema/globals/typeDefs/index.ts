import { Query } from './entrypoint';
import scalars from './scalars';

const globals = [Query, scalars];

export default globals;
