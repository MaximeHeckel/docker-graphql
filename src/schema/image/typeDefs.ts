import { gql } from "apollo-server";

const typeDefs = gql`
  extend type Query {
    images: [Image!]!
    image(name: String!): ImageInspect!
  }

  type Image {
    Id: ID!
    ParentId: ID
    RepoTags: [String]
    RepoDigests: [String]
    Created: Int!
    Size: Int!
    VirtualSize: Int!
    SharedSize: Int
    Labels: JSON
    Containers: Int
  }

  type GraphDriverType {
    Name: String
    Data: JSON
  }

  type ImageConfigType {
    Image: ID!
    NetworkDisabled: Boolean
    OnBuild: [String]!
    StdinOnce: Boolean
    PublishService: String
    AttachStdin: Boolean
    OpenStdin: Boolean
    Domainname: String
    AttachStdout: Boolean
    Tty: Boolean
    Hostname: String!
    Cmd: [String]
    Env: [String]
    Labels: JSON
    MacAddress: String
    AttachStderr: Boolean
    WorkingDir: String
    User: String
  }

  type RootFSType {
    Type: String
    Layers: [ID!]!
  }

  type ImageInspect {
    Id: ID!
    Container: ID
    Comment: String
    OS: String
    Architecture: String
    Parent: ID
    ContainerConfig: ContainerConfigType
    DockerVersion: String!
    VirtualSize: Int!
    Size: Int!
    Author: String
    Created: String!
    GraphDriver: GraphDriverType
    RepoDigests: [String]
    RepoTags: [String]
    Config: ImageConfigType
    RootFS: RootFSType
  }
`;

// TODO: Check for Architecture and RootFS Type enum

export default typeDefs;
