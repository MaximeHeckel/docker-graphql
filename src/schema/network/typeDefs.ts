import { gql } from "apollo-server";

const typeDefs = gql`
  extend type Query {
    networks: [Network!]!
    network(id: ID!): NetworkInspect!
  }

  enum NetworkScopeEnum {
    swarm
    global
    local
  }

  type IPAMConfigType {
    Subnet: String!
    Gateway: String!
  }

  type IPAMType {
    Driver: String!
    Config: [IPAMConfigType]
    Options: JSON
  }

  type Network {
    Name: String
    Id: ID!
    Created: String!
    Scope: NetworkScopeEnum!
    Driver: String!
    EnableIPv6: Boolean
    Internal: Boolean
    Attachable: Boolean
    Ingress: Boolean
    IPAM: IPAMType
    Options: JSON
  }

  type NetworkInspect {
    Name: String
    Id: ID!
    Created: String!
    Scope: NetworkScopeEnum!
    Driver: String!
    EnableIPv6: Boolean
    Internal: Boolean
    Attachable: Boolean
    Ingress: Boolean
    IPAM: IPAMType
    Options: JSON
    Labels: JSON
    Containers: JSON
  }
`;

export default typeDefs;
