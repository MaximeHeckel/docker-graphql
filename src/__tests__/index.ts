import { ApolloServer } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import { gql } from 'apollo-server';
import { baseURL } from '../constants';
import typeDefs from '../schema/typeDefs';

const mocks = {}; // Mock any type in this object

const resolvers = {
  Query: {
    services: () => {
      body: [
        {
          ID: '9mnpnzenvg8p8tdbtq4wvbkcz',
          Version: {
            Index: 19,
          },
          CreatedAt: '2016-06-07T21:05:51.880065305Z',
          UpdatedAt: '2016-06-07T21:07:29.962229872Z',
          Spec: {
            Name: 'hopeful_cori',
            TaskTemplate: {
              ContainerSpec: {
                Image: 'redis',
              },
              Resources: {
                Limits: {},
                Reservations: {},
              },
              RestartPolicy: {
                Condition: 'any',
                MaxAttempts: 0,
              },
              Placement: {},
              ForceUpdate: 0,
            },
            Mode: {
              Replicated: {
                Replicas: 1,
              },
            },
            UpdateConfig: {
              Parallelism: 1,
              Delay: 1000000000,
              FailureAction: 'pause',
              Monitor: 15000000000,
              MaxFailureRatio: 0.15,
            },
            RollbackConfig: {
              Parallelism: 1,
              Delay: 1000000000,
              FailureAction: 'pause',
              Monitor: 15000000000,
              MaxFailureRatio: 0.15,
            },
            EndpointSpec: {
              Mode: 'vip',
              Ports: [
                {
                  Protocol: 'tcp',
                  TargetPort: 6379,
                  PublishedPort: 30001,
                },
              ],
            },
          },
          Endpoint: {
            Spec: {
              Mode: 'vip',
              Ports: [
                {
                  Protocol: 'tcp',
                  TargetPort: 6379,
                  PublishedPort: 30001,
                },
              ],
            },
            Ports: [
              {
                Protocol: 'tcp',
                TargetPort: 6379,
                PublishedPort: 30001,
              },
            ],
            VirtualIPs: [
              {
                NetworkID: '4qvuz4ko70xaltuqbt8956gd1',
                Addr: '10.255.0.2/16',
              },
              {
                NetworkID: '4qvuz4ko70xaltuqbt8956gd1',
                Addr: '10.255.0.3/16',
              },
            ],
          },
        },
      ];
    },
  },
};

const constructTestServer = ({ context }) => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    mocks,
  });

  return { server };
};

const GET_SERVICES = gql`
  query {
    services {
      ID
      Spec {
        Name
      }
    }
  }
`;

describe('Integration tests', () => {
  it('Fetches the list of services', async () => {
    const { server } = constructTestServer({
      context: () => ({ baseURL }),
    });

    const { query } = createTestClient(server);

    const res = await query({ query: GET_SERVICES });
    expect(res.data.services).toMatchSnapshot();
  });
});
