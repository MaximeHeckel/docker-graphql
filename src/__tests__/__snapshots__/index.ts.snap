// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests Fetches the list of services 1`] = `
Array [
  Object {
    "ID": "e95ea8fe-2acf-4194-b219-0899d8951e51",
    "Spec": Object {
      "Name": "Hello World",
    },
  },
  Object {
    "ID": "b682e41a-4f32-49ae-9acc-48f01a5c6e1e",
    "Spec": Object {
      "Name": "Hello World",
    },
  },
]
`;
